### DISCUSSION TITLE
Advanced: Windows installation menu with autoboot and multiple images

### DISCUSSION THREAD

**Discussion by nwpr** [antonym/netboot.xyz#757] [Show and tell]
I'd like to share our setup for installing Windows by just choosing the needed image!
The custom ipxe files are hosted on Github. (Except for the one loading the variables.)
It is structured so that we have the possibility to add other menus if we want to. I used the windows.ipxe by netboot.xyz as template.

Note: This aims to be a guide for advanced users, so I won't explain every detail.

All `$$value$$` variables have to be replaced with the correct values.

First of all you need to fork the netboot.xyz-custom repository on Github and edit the custom.ipxe file. Here's what the file looks like:

custom.ipxe
```
#!ipxe

goto ${menu} ||

:custom_menu
menu $$username$$ custom menu
item --gap OS:
item windows ${space} Windows
choose menu || goto custom_exit
echo ${cls}
goto ${menu} ||
goto change_menu

:change_menu
chain --autofree https://raw.githubusercontent.com/${github_user}/netboot.xyz-custom/master/${menu}.ipxe || goto error
goto custom_menu

:custom_exit
exit 0
```

Afterwards you can create a file called windows.ipxe with the following content:

windows.ipxe
```
#!ipxe

# Microsoft Windows
# https://www.microsoft.com

set winpe_arch x64
set win_image Win10_20H2_English_x64
set win_base_url $$win_base_url$$
goto ${menu} ||

:windows
set os Microsoft Windows
clear win_version
menu ${os} 
item --gap Installers
item win_install ${space} Load ${os} Installer...
item --gap Options:
item image_set ${space} Image [ ${win_image} ]
item pe_arch_set ${space} Architecture [ ${winpe_arch} ]
item url_set ${space} Base URL [ ${win_base_url} ]
choose win_version || goto windows_exit
goto ${win_version}

:image_set
menu Image
item Win10_1909_English_x64 Win10_1909_English_x64
item Win10_20H2_English_x64 Win10_20H2_English_x64
choose win_image && goto windows

:pe_arch_set
iseq ${winpe_arch} x64 && set winpe_arch x86 || set winpe_arch x64
goto windows

:url_set
echo Set the HTTP URL of an extracted Windows ISO without the trailing slash:
echo e.g. http://www.mydomain.com/windows
echo
echo -n URL: ${} && read win_base_url
echo
echo netboot.xyz will attempt to load the following files:
echo ${win_base_url}/${winpe_arch}/bootmgr
echo ${win_base_url}/${winpe_arch}/bootmgr.efi
echo ${win_base_url}/${winpe_arch}/boot/bcd
echo ${win_base_url}/${winpe_arch}/boot/boot.sdi
echo ${win_base_url}/${winpe_arch}/sources/boot.wim
echo ${win_base_url}/configs/${win_image}/install.bat
echo ${win_base_url}/configs/configure.bat
echo ${win_base_url}/configs/winpeshl.ini
echo
prompt Press any key to return to Windows Menu...
goto windows

:win_install
isset ${win_base_url} && goto boot || echo URL not set... && goto url_set

:boot
imgfree
kernel http://${boot_domain}/wimboot
initrd ${win_base_url}/configs/${win_image}/install.bat install.bat
initrd ${win_base_url}/configs/configure.bat configure.bat
initrd ${win_base_url}/configs/winpeshl.ini winpeshl.ini
initrd -n bootmgr     ${win_base_url}/${winpe_arch}/bootmgr       bootmgr ||
initrd -n bootmgr.efi ${win_base_url}/${winpe_arch}/bootmgr.efi   bootmgr.efi ||      
initrd -n bcd         ${win_base_url}/${winpe_arch}/boot/bcd      bcd ||
initrd -n bcd         ${win_base_url}/${winpe_arch}/Boot/BCD      bcd ||
initrd -n boot.sdi    ${win_base_url}/${winpe_arch}/boot/boot.sdi boot.sdi ||
initrd -n boot.sdi    ${win_base_url}/${winpe_arch}/Boot/boot.sdi boot.sdi ||
initrd -n boot.wim    ${win_base_url}/${winpe_arch}/sources/boot.wim boot.wim
boot

:windows_exit
exit 0
```

What we are doing here is booting WinPE with predefined config files, which change the boot behavior so we don't need to enter any commands afterwards.

Now we have to create the local files required.

Download WinPE/generate the image and download Windows ISOs.
Setup a web server at $$win_base_url$$ hosting the WinPE extracted image.
Setup an SMB share with Windows ISOs extracted to a directory there.

Now we have to additionally create these folders and files on the web server $$win_base_url$$:
```
- configs/
   - configure.bat
   - winpeshl.ini
   - $Win10_20H2_English_x64$/
      - install.bat
   - $Win10_1909_English_x64$/
      - install.bat
- x64/
   - *WinPE_x64*
- x86/
   - *WinPE_x86*
```
`configure.bat` and `winpeshl.ini` are always the same files. But the `install.bat` changes for each image.

configure.bat
```
@echo off
Wpeutil InitializeNetwork
:START
ping -n 1 $$NET_SHARE_HOSTNAME$$
if errorlevel 1 GOTO START
net use m: $$NET_SHARE$$ /user:$$USERNAME$$ $$PASSWORD$$
install.bat
```
(It often happens that WinPE doesn't wait for the network to come up. So we are pinging the server until it is reachable.)

winpeshl.ini
```
[LaunchApp]
AppPath = configure.bat
```

Example of install.bat
```
m:\Win10_20H2_English_x64\setup.exe
```

To use the custom menu just set your Github user in the utilites menu and choose your new custom menu!
I hope this setup helps some people to save time, as we needed many hours of trail and error until we had this setup working properly ;)

**Comment by samip5** [comment]
Oh, so that's how you do it the correct way. Thank you. I were wondering how I can get it to boot directly to the setup. :)

**Comment by noagenda33** [comment]
Is there something ready-made for the beginner like me? I just got MeshCentral with Intel AMT working and discovered this netbook.xyz.iso. Seems like a great tool but am lost on getting Windows 10 to install. I have a Windows file server I can expose to the internet. I have created a WinPE.iso. But thats as far as I can get. Any help for a beginner is appreciated.

**Reply by uzairali001** [reply]
Do let me know when you found something, I'm also looking for an easy guide

**Reply by uzairali001** [reply]
Btw do I really need WinPE? isn't it a Windows Preinstallation Environment so why do I need it for windows10 setup?

**Reply by nodiaque** [reply]
To answer the question. I just tried booting directly the wim through pxe. Problem is once booted, it can't find the file for the setup.exe since it's not loaded. You could setup stuff for it but that way work well

**Comment by itsnanobug** [comment]
Hello,
I've tried following your step, but I'm having issues...
I'm not sure of it all, so I might be too detailed or not detailed enough, so I'll try my be to be too detailed.

In the x64 and x84 folders it says:
` - *WinPE_x64*`
 For both, what are those?

It's running in a container on Unraid where I made a SMB share called netbootxyz that guests have read access to.

Current folder structure:
```
netbootxyz
	assets
		WinPE
			config
				$Win10
				$Win11
				configure.bat
				winpeshl.ini
			x64
			x86
	windows
		10
			boot
			efi
			sources
			support
			autorun.inf
			bootmgr
			bootmgr.efi
			setup.exe
		11
			boot
			efi
			sources
			support
			autorun.inf
			bootmgr
			bootmgr.efi
			setup.exe
```

The $Win10_1909_English_x64$ install.bat file:
`m:\10.0.10.50\netbootxyz\windows\10`

The Win10_20H2_English_x64$ install.bat file:
`m:\10.0.10.50\netbootxyz\windows\11`

I know the don't correspond correctly. At first I just want it to work.

At first, this menu comes up:
![image](https://github.com/netbootxyz/netboot.xyz/assets/64610851/43f5229f-1db0-404b-80ff-46a6b21f72b8)

I click enter and it goes to the default netboot.xyz menu.
![image](https://github.com/netbootxyz/netboot.xyz/assets/64610851/8c7b1b54-0870-49bc-b892-17d6dd6fc308)

I assume that's not how it's supposed to work.

Also regarding the iso files.
I couldn't exactly figure out how to do it.
But I've mounted the iso for each Windows version, then all the folders and files has been copied to the locations mentioned above.
Is this correct?

I've only posted the changes that I've made to the files, everything else is default and copy pasted.

Where am I doing this wrong?

**Reply by diaverso** [reply]
Did you find how to make it work?

**Comment by megapearl** [comment]
Can I pass variables from netboot xyz to winpeshl.ini and install.bat? like architecture and mounted windows installation directory?

**Reply by davegermiquet** [reply]
Yes I'm trying to do the same thing with netbootxyz I want to pass the mac address and host name to the install.bat file any ideas how to do that?

**Comment by Eurodocsis** [comment]
A bit delayed ->  Thank you But got this working !   Needed to add some drives to boot.wim to fully get the network going Thanks
